brightwind/analyse/correlation.py

```diff
@@ -949,53 +952,85 @@ def plot(self):
+            # then the first and the last sectors are taken into account for deriving the veer as for code below.
+            if (0 in self.direction_bin_array) and (360 in self.direction_bin_array):
+                sec_veers.insert(0, self._linear_interpolation(0 - (360 - veer_bins[-1]), veer_bins[0],
+                                                               sec_veers[-1], sec_veers[0], 0))
+                sec_veers.append(self._linear_interpolation(veer_bins[-1], 360 + veer_bins[0],
+                                                            sec_veers[-1], sec_veers[1], 360))
+                veer_bins.insert(0, 0)
+                veer_bins.append(360)
+
+        # The veer correction is derived linear interpolating the veer between two mid-points of near sectors.
+        adjustment = x_dir.rename('adjustment').copy() * np.nan
+        for i in range(1, len(veer_bins)):
+            logic_sect = self._get_logic_dir_sector(ref_dir=x_dir,
+                                                    sector_min=veer_bins[i - 1],
+                                                    sector_max=veer_bins[i])
+
+            adjustment[logic_sect] = self._linear_interpolation(veer_bins[i - 1], veer_bins[i],
+                                                                sec_veers[i - 1], sec_veers[i],
+                                                                x_dir[logic_sect])
+        return offset_wind_direction(x_dir, adjustment).sort_index()
 
     def _predict(self, x_spd, x_dir):
-        x = pd.concat([x_spd.dropna().rename('spd'),
-                       _binned_direction_series(x_dir.dropna(), self.sectors,
+        x = pd.concat([x_spd.rename('spd'),
+                       _binned_direction_series(x_dir, self.sectors,
                                                 direction_bin_array=self.direction_bin_array).rename('ref_dir_bin')],
-                      axis=1, join='inner')
-        prediction = pd.DataFrame()
-        first = True
+                      axis=1, join='inner').dropna()
+        prediction = pd.Series().rename('spd')
         for sector, data in x.groupby(['ref_dir_bin']):
-            if first is True:
-                first = False
-                prediction = self.speed_model[sector].sector_predict(data['spd'])
+            if sector in list(self.speed_model.keys()):
+                prediction_spd = self.speed_model[sector].sector_predict(data['spd'])
             else:
-                prediction = pd.concat([prediction, self.speed_model[sector].sector_predict(data['spd'])], axis=0)
+                prediction_spd = data['spd'] * np.nan
+
+            prediction = pd.concat([prediction, prediction_spd], axis=0)
 
         return prediction.sort_index()
 
     def synthesize(self, input_spd=None, input_dir=None):
         # This will give erroneous result when the averaging period is not a whole number such that ref and target does
-        # bot get aligned -Inder
+        # both get aligned -Inder
```

Comment:
Is Inder's comment still valid since I refactored the average_data_by_period function?

Reply:
solved